### PR TITLE
HZN-1256: Add support for forwading events, alarms and nodes to Kafka topics

### PR DESCRIPTION
**Pull request by j-white**
JIRA: https://issues.opennms.org/browse/HZN-1256

Here we make it possible for forward events, alarms and nodes formatted as Google Protocol Buffers (GPBs) into Kafka topics.

We also extend the Alarm API to support registering for callbacks on alarm life-cycle events (similar to how this achieved in bsmd) and leverage this for forwarding alarms to Kafka when they are changed.

### CODE REVIEW

**Comment by cgorantla** [comment]
Verified on my side. Looks good to me :+1:

File: opennms-doc/guide-admin/src/asciidoc/text/kafka-producer/kafka-producer.adoc [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+The Kafka Producer will also perform periodic synchronization tasks to ensure that the contents of the Kafka topic reflect the current state of alarms in the _{opennms-product-name}_ database.
+
+By default, all alarms (and subsequent updates) are forwarded to a topic named `alarms`.
+
+The name of the topic used can be configured, and an optional filtering expression can be set to help control which alarms are sent to the topic.
+
+==== Nodes
+
+If an event or alarm being forwarded reference a node, then the corresponding node is also forwarded.
+The records are keyed by "node criteria" (see bellow) and contain a GPB encoded model of the alarm.
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/kafka-producer/kafka-producer.adoc]: > contain a GPB encoded model of the alarm

Question: The records contain a GPB encoded alarm or a node?

File: opennms-doc/guide-admin/src/asciidoc/text/kafka-producer/kafka-producer.adoc [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+admin@opennms()> config:update
+----
+
+Next, install the `opennms-kafka-producer` feature from that same shell using:
+[source]
+----
+admin@opennms()> feature:install opennms-kafka-producer
+----
+
+In order to ensure that the feature continues to be installed as subsequent restarts, add `opennms-kafka-producer` to the `featuresBoot` property in the `${OPENNMS_HOME}/etc/org.apache.karaf.features.cfg`
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/kafka-producer/kafka-producer.adoc]: Is there a way to add this to a dedicated file which can be easily updated? As described here, this makes updating and maintaining opennms harder and error prune.

File: opennms-doc/guide-admin/src/asciidoc/text/kafka-producer/kafka-producer.adoc [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+The records are keyed by event _UEI_ and contain a _GPB_ encoded model of the event.
+
+By default, all events are forwarded to a topic named `events`.
+
+The name of the topic used can be configured, and an optional filtering expression can be set to help control which events are sent to the topic.
+
+==== Alarms
+
+The _Kafka Producer_ listens for changes made to the current set of alarms and forwards the resulting alarms to a _Kafka_ topic.
+The records are keyed by alarm reduction key and contain a _GPB_ encoded model of the alarm.
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/kafka-producer/kafka-producer.adoc]: > ... contain a GPB encoded model of the alarm

Question: The records contain a GPB encoded alarm or a node?

File: opennms-doc/guide-admin/src/asciidoc/text/kafka-producer/kafka-producer.adoc [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+----
+
+Next, install the `opennms-kafka-producer` feature from that same shell using:
+
+[source]
+----
+admin@opennms()> feature:install opennms-kafka-producer
+----
+
+In order to ensure that the feature continues to be installed as subsequent restarts, add `opennms-kafka-producer` to the `featuresBoot` property in the `${OPENNMS_HOME}/etc/org.apache.karaf.features.cfg`.
```
Comment (indigo423) [inline opennms-doc/guide-admin/src/asciidoc/text/kafka-producer/kafka-producer.adoc]: Just notice and non-critical question. Is there a way to add this to a dedicated file which can be easily updated? As described here, this makes updating and maintaining opennms harder and error prune.

File: opennms-doc/guide-admin/src/asciidoc/text/kafka-producer/kafka-producer.adoc (hunk 3 above)
Comment (j-white) [reply opennms-doc/guide-admin/src/asciidoc/text/kafka-producer/kafka-producer.adoc]: In the alarms topic, it is a GPB encoded alarm.

File: opennms-doc/guide-admin/src/asciidoc/text/kafka-producer/kafka-producer.adoc (hunk 4 above)
Comment (j-white) [reply opennms-doc/guide-admin/src/asciidoc/text/kafka-producer/kafka-producer.adoc]: Agreed, this same point actually came up earlier today too.

In Minion we can do this in the `featuresBoot.d` folder, we'll be able to do the same in OpenNMS soon (dependent of some refactoring which is being performed as part of HZN-1284).

**Comment by j-white** [comment]
The feature was also validated in a customer environment. Merging for inclusion in 21.1.0.

**Pull request merged**